### PR TITLE
Crm 16259

### PR DESCRIPTION
**Pull request by pradpnayak**

### CODE REVIEW

**Comment by davecivicrm** [comment]
@pradpnayak The upgrade changes need to be moved from alpha1 file to either alpha5 or beta1 (I'll check w/ @kurund to confirm which).

@pradpnayak Also the two style warnings led me to find that there are debug statements on those lines which I assume are not intended to be merged in - Contribution.php lines 4476 and 4480.

**Comment by davecivicrm** [comment]
@pradpnayak Upgrade code definitely needs to be moved to beta1 if we want this merged for next Wednesday's release.

**Comment by davecivicrm** [comment]
@pradpnayak There are 5 failures in syntaxConformanceTest for the Payment API. These need to be addressed before we can merge this.
https://test.civicrm.org/job/CiviCRM-Core-PR/6166/

**Comment by kurund** [comment]
jenkins, test this please

File: api/v3/Payment.php [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    $ftParams = array(
+      'id' => array('IN' => $eftIds),
+      'is_payment' => 1,
+    );
+    $financialTrxn = civicrm_api3('FinancialTrxn', 'get', $ftParams);
+    foreach ($financialTrxn['values'] as &$values) {
+      $values['contribution_id'] = $map[$values['id']];
+    }
+  }
+  return civicrm_api3_create_success($financialTrxn['values'], $params, 'Payment', 'cancel');
```
Comment (monishdeb) [inline api/v3/Payment.php]: api_v3_SyntaxConformanceTest::testSimple_get
api_v3_SyntaxConformanceTest::testGetfieldsHasTitle
api_v3_SyntaxConformanceTest::testSqlOperators.testSqlOperators 
all 3 failing due to 'Undefined index: values'
 $financialTrxn['values'] should be changed to CRM_Utils_Array::value('values', $financialTrxn, $financialTrxn) in return

**Comment by monishdeb** [comment]
api_v3_SyntaxConformanceTest::testCreateSingleValueAlter with data set #43 ('Payment')
api_v3_SyntaxConformanceTest::testGetfieldsHasTitle
Failure : 'Failed asserting that a NULL is not empty'.
Occuring due to Payment entity isn't calling any corrosponding DAO/BAO class as a result _civicrm_api3_get_BAO('Payment') returns NULL (or we should simple skip that for now via adding 'Payment' entity in https://github.com/civicrm/civicrm-core/blob/master/tests/phpunit/api/v3/SyntaxConformanceTest.php#L96) and metadata for 'Payment' entity fields is empty for getfield action 'get'

**Comment by pradpnayak** [comment]
jenkins, test this please

**Comment by JoeMurray** [comment]
From what I can see in terms of files, the errors are unrelated to our patch.

**Comment by kurund** [comment]
@pradpnayak you need to fix the warnings reported: https://test.civicrm.org/job/CiviCRM-Core-PR/6769/ Also this PR will need to be rebased.

**Comment by pradpnayak** [comment]
jenkins, test this please

**Comment by pradpnayak** [comment]
Rebased PR but details show up as 0 warnings and errors.

**Comment by pradpnayak** [comment]
I'm wondering why jenkins is failing here since details on the merge build suggest 0 warnings have been found.

**Comment by JoeMurray** [comment]
Where is link to new Jenkins test results? I see 65 warnings on old test results.

**Comment by pradpnayak** [comment]
https://test.civicrm.org/job/CiviCRM-Core-PR/6852/

**Comment by davecivicrm** [comment]
@pradpnayak Something up with test infrastructure. Tim is reviewing. I'll run syntax conform tests locally if not fixed soon so we can merge today.

**Comment by totten** [comment]
jenkins, test this please

**Comment by davecivicrm** [comment]
@pradpnayak  Tests have run properly now. There are 6 failures reported at https://test.civicrm.org/job/CiviCRM-Core-PR/6859/

One is unrelated. However, these 5 are related to this PR and need attention:
api_v3_SyntaxConformanceTest::testSimple_get.testSimple_get with data set #77
api_v3_SyntaxConformanceTest::testAcceptsOnlyID_get.testAcceptsOnlyID_get with data set #77
api_v3_SyntaxConformanceTest::testSqlOperators.testSqlOperators with data set #76
api_v3_SyntaxConformanceTest::testCreateSingleValueAlter.testCreateSingleValueAlter with data set #43
api_v3_SyntaxConformanceTest::testGetfieldsHasTitle.testGetfieldsHasTitle with data set #80

**Comment by davecivicrm** [comment]
Schema change only merged at https://github.com/civicrm/civicrm-core/pull/7362

**Comment by pradpnayak** [comment]
jenkins, test this please

**Comment by eileenmcnaughton** [comment]
I added a new PR specifically on excluding payment from unit tests - - wait until that is merged before rebasing https://github.com/civicrm/civicrm-core/pull/7406

File: CRM/Contribute/BAO/Contribution.php [hunk 2]
```diff
@@ -2959,6 +2959,10 @@ public static function recordFinancialAccounts(&$params, $financialTrxnValues =
         $balanceTrxnParams['status_id'] = $statusId;
         $balanceTrxnParams['payment_instrument_id'] = $params['contribution']->payment_instrument_id;
         $balanceTrxnParams['check_number'] = CRM_Utils_Array::value('check_number', $params);
+        if (!empty($balanceTrxnParams['from_financial_account_id']) &&
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution.php]: So, I would do a separate PR with just the changes to set 'is_payment' (generally with a BAO level test at the same time)

File: CRM/Contribute/BAO/Contribution.php [hunk 3]
```diff
@@ -3014,6 +3018,9 @@ public static function recordFinancialAccounts(&$params, $financialTrxnValues =
         'payment_instrument_id' => $params['contribution']->payment_instrument_id,
         'check_number' => CRM_Utils_Array::value('check_number', $params),
       );
+      if (!in_array(CRM_Utils_Array::value('contribution_status_id', $params), $pendingStatus)) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution.php]: ie - this row too in the above comment

File: CRM/Contribute/BAO/Contribution.php [hunk 4]
```diff
@@ -3626,7 +3639,9 @@ public static function recordAdditionalPayment($contributionId, $trxnsData, $pay
       // the reason because it performs 'status change' related code execution for financial records
       // which in 'Pending Refund' => 'Completed' is not useful, instead specific financial record updates
       // are coded below i.e. just updating financial_item status to 'Paid'
-      $contributionDetails = CRM_Core_DAO::setFieldValue('CRM_Contribute_BAO_Contribution', $contributionId, 'contribution_status_id', $statusId);
+      if ($updateStatus) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution.php]: Once you have merged the rows from this line up (preferably with tests)  should not get many more merge conflicts

File: CRM/Contribute/BAO/Contribution.php [hunk 5]
```diff
@@ -4492,4 +4507,113 @@ public function loadRelatedMembershipObjects(&$ids) {
     }
   }
 
+  public static function recordPartialPayment($contribution, $params) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution.php]: (NB comment block needed)

**Comment by JoeMurray** [comment]
Pradeep/Edsel, here's our path forward on this after discussion with Eileen and Dave:
1. Take out the UI stuff for UserDashboard including change to tpl that inserts the new tab.
2. Make small PRs for the BAO stuff, with each PR having a function and associated test(s). NB: It's essential to follow coding standards in terms of including a comment block for each function for BAO.
3. After the BAO PRs are committed, then the API code should be able to go in smoothly with associated tests.

**Comment by eileenmcnaughton** [comment]
I've merged the other one. NB I've I were doing this I would apply it over the top of of origin/master & then recommit it in chunks - since this history is pretty long - these are the steps I would take to split it up per Joe's instructions. 

git fetch -f origin
git checkout master 
git pull origin master 
(ie. make sure you are checked out on a clean master & the remote branches are all up to date as well)
git pull jma CRM-16259
git reset origin/master

git add -p

& then add the smallest change that 'stands alone' & commit it & do a PR of that. I try not to keep my PRs small & if the code is not co-dependent then put it in separate PRs

NB - I know you have your own workflow - so don't feel under any obligation to take note of mine :-) but maybe it will help someone

**Comment by JoeMurray** [comment]
Many thanks, Eileen!

**Comment by pradpnayak** [comment]
closing in favour of 
https://github.com/civicrm/civicrm-core/pull/7583
https://github.com/civicrm/civicrm-core/pull/7476
https://github.com/civicrm/civicrm-core/pull/7474
https://github.com/civicrm/civicrm-core/pull/7477
https://github.com/civicrm/civicrm-core/pull/7475
https://github.com/civicrm/civicrm-core/pull/7473
